pkg/commands/docker_source.go

```diff
@@ -0,0 +1,225 @@
+package commands
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	build "github.com/Azure/acr-builder/pkg"
+	"github.com/Azure/acr-builder/pkg/constants"
+	dockerbuild "github.com/docker/cli/cli/command/image/build"
+	"github.com/docker/docker/builder/remotecontext/git"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/docker/docker/pkg/urlutil"
+	"github.com/pkg/errors"
+)
+
+// NewDockerSource creates a new passthrough source
+func NewDockerSource(context, dockerfile string) build.Source {
+	return &DockerSource{
+		context:    context,
+		dockerfile: dockerfile,
+	}
+}
+
+type dockerSourceType int
+
+const (
+	dockerSourceLocal dockerSourceType = iota
+	dockerSourceGit
+	dockerSourceDockerfile
+	dockerSourceArchive
+)
+
+// DockerSource is a source we can pass directly into docker
+type DockerSource struct {
+	// context is the literal string representing docker build context. It can be a directory, archive url, git url or "-"
+	context    string
+	dockerfile string
+	tracker    *DirectoryTracker
+	gitHeadRev string
+}
+
+// Obtain obtains the source
+func (s *DockerSource) Obtain(runner build.Runner) error {
+	runContext := runner.GetContext()
+	context := runContext.Expand(s.context)
+	dockerfile := runContext.Expand(s.dockerfile)
+	if context == constants.FromStdin && dockerfile == constants.FromStdin {
+		return fmt.Errorf("invalid argument: can't use stdin for both build context and dockerfile")
+	}
+	sourceType, workingDir, err := s.ensureContext(runner, context, dockerfile)
+	if err != nil {
+		return err
+	}
+	if workingDir != "" {
+		s.tracker, err = ChdirWithTracking(runner, workingDir)
+		if err != nil {
+			return err
+		}
+	}
+	if sourceType == dockerSourceGit {
+		sha, queryErr := runner.QueryCmd("git", []string{"rev-parse", "--verify", "HEAD"})
+		if queryErr != nil {
+			fmt.Fprintf(os.Stderr, "Error querying for git head rev: %s, output will not contain git head revision SHA", queryErr)
```

Comment:
Why not just use `logrus` everywhere, it's a mixed bag right now

Reply:
Changed